### PR TITLE
[FEATURE] Rasterize2D function for MSExperiment

### PR DESCRIPTION
**Pull request by jpfeuffer**
- **[FEATURE] Add rasterize2D method for MSExperiment**
- **[FIX] Qualify RasterAggregation enum in Cython addon**
- **[BUILD] Add build module to dependency-groups.build**
- **fix macos dependency script to remove superfluous rpaths after copying (avoids issues with finding multiple dylibs at runtime)**
- **create venv in build folder and clarify pyopenms README**
- **add rasterize2D function. No IM yet.**
- **fix annoying warnings in SpectraMerger**

## Release Notes

* **New Features**
  * Added spectral rasterization to 2D RT × m/z grids (SUM/MAX) and ion-mobility frame rasterization.
  * Exposed rasterization APIs to Python (pyOpenMS) and added optional pyOpenMS venv Python version setting.

* **Bug Fixes**
  * Fixed unsafe formatting usage to prevent potential buffer overflow.
  * Improved macOS RPATH management for bundled binaries.

* **Documentation**
  * Updated pyOpenMS build docs with uv/venv guidance.

* **Tests**
  * Added comprehensive unit and Python tests for rasterization features.

### CODE REVIEW

**Comment by jpfeuffer** [comment]
@timosachsenberg Maybe add IM functionality. Not sure what would need to be done for this. Would one want 3D? Or just filter for an IM range and still collapse.

**Comment by timosachsenberg** [comment]
I think we want to collapse IM if we want to plot 2D mz/rt maps
And also rasterize frames if we want to plot m/z IM maps

**Comment by jpfeuffer** [comment]
TODOs:
- Need to find OpenMP workarounds for stupid ugly Windows
- Fix tests and add better real ones. GPT5 is too stupid to calculate things.
- Add IM testing and potentially functionality

**Comment by timosachsenberg** [comment]
can do

**Comment by jpfeuffer** [comment]
I saw the tests assume different boundary conditions in the range. E.g. if to include <= or < on the bin boundaries.

File: src/pyOpenMS/CMakeLists.txt [hunk 1]
```diff
@@ -523,7 +565,7 @@ WITH_PARQUET = ${_PY_WITH_PARQUET}
     if(APPLE)
       # macOS: Fix dependencies and codesign after copying
       add_custom_target(pyopenms_fix_deps ALL
-        COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/../../cmake/MacOSX/fix_dependencies.rb -l "${PYOPENMS_BUILD_DIR}/pyopenms" -e "@rpath/" -f -v
+        COMMAND ${PROJECT_SOURCE_DIR}/mac_fix_dependencies.rb -l "${PYOPENMS_BUILD_DIR}/pyopenms" -e "@rpath/" -r "@loader_path/" -f -v
         COMMAND codesign --force --sign - ${PYOPENMS_BUILD_DIR}/pyopenms/Qt* || true
         COMMAND codesign --force --sign - ${PYOPENMS_BUILD_DIR}/pyopenms/*.dylib || true
```
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/pyOpenMS/CMakeLists.txt:570]: bullsht. there is a symlink

File: cmake/MacOSX/fix_dependencies.rb [hunk 2]
```diff
@@ -90,6 +91,25 @@ def fixId(target, libname)
   `#{$install_name_tool} -id #{$executableId}#{libname} #{target}`
 end
 
+###############################################################################
+def adjustRpaths(target, newpaths)
+  return if newpaths.nil?
+  # list existing rpaths
+  existing = %x[otool -l #{target} | grep -A2 LC_RPATH | grep path | sed -n "s/^.*path\\s*\\(.*\\)(offset.*\\$/\\1/p"].split(/\n/).map(&:strip)
```
Comment (timosachsenberg) [inline cmake/MacOSX/fix_dependencies.rb:98]: what use case does this serve?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -37,24 +41,7 @@ if(PYOPENMS)
...
-    if(NOT _uv_install_result EQUAL 0)
-      message(WARNING "Failed to install pyOpenMS build dependencies with uv: ${_uv_install_error}")
-    endif()
-    if(WIN32)
-      set(Python_EXECUTABLE "${CMAKE_CURRENT_LIST_DIR}/pyOpenMS/.venv/Scripts/python.exe")
-    else()
-      set(Python_EXECUTABLE "${CMAKE_CURRENT_LIST_DIR}/pyOpenMS/.venv/bin/python")
-    endif()
-  elseif(NOT DEFINED Python_EXECUTABLE)
-    set(Python_EXECUTABLE "python")
```
Comment (timosachsenberg) [inline src/CMakeLists.txt:57]: why is this not needed (anymore)?

Comment (jpfeuffer) [reply src/CMakeLists.txt:57]: this is in pyopenms/CMakeLists now

File: src/openms/include/OpenMS/KERNEL/MSExperiment.h [hunk 4]
```diff
@@ -408,6 +408,74 @@ namespace OpenMS
...
+    * @note This method is thread-safe and uses per-thread accumulation buffers to avoid
+     *       contention, then merges results at the end.
+     *
+     * Example usage with numpy (via pyOpenMS):
+     * @code
+     * import numpy as np
+     * rt_bins, mz_bins = 800, 600
+     * output = np.zeros((mz_bins, rt_bins), dtype=np.float32)
+     * exp.rasterize2D(output, rt_bins, mz_bins, min_rt, max_rt, min_mz, max_mz, 1)
+     * @endcode
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/KERNEL/MSExperiment.h:466]: I *think* this could also work with nanobind. Can you confirm?

File: src/CMakeLists.txt (hunk 3 above)
Comment (jpfeuffer) [reply src/CMakeLists.txt:57]: It depends on the copy step of the pyproject.toml to the build folder since I am creating the venv in the build folder now.
Not sure what is the best design but I think the venv should definitely be in the build folder. In theory you could have multiple build folders with different python versions, etc.

File: src/openms/include/OpenMS/PROCESSING/SPECTRAMERGING/SpectraMerger.h [hunk 5]
```diff
@@ -731,7 +731,7 @@ namespace OpenMS
       }
 
       char buffer[200];
-      sprintf(buffer, "%d/%d (%.2f %%) of blocked spectra", (int)count_peaks_aligned,
+      std::snprintf(buffer, sizeof(buffer), "%d/%d (%.2f %%) of blocked spectra", (int)count_peaks_aligned,
               (int)count_peaks_overall, float(count_peaks_aligned) / float(count_peaks_overall) * 100.);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/PROCESSING/SPECTRAMERGING/SpectraMerger.h:735]: wth std::format and << would work just as well

File: src/openms/include/OpenMS/KERNEL/MSExperiment.h (hunk 4 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/KERNEL/MSExperiment.h:466]: Yes, as long as you dont use vector in the C++ interface you can probably accept these kind of things with nanobind.
Ideally you would accept a span or even better, an Eigen::Map which is a span with Matrix operations.

File: src/openms/include/OpenMS/PROCESSING/SPECTRAMERGING/SpectraMerger.h (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/PROCESSING/SPECTRAMERGING/SpectraMerger.h:735]: most likely. this was the closest replacement

File: src/openms/include/OpenMS/KERNEL/MSExperiment.h (hunk 4 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/KERNEL/MSExperiment.h:466]: Converting to vector in the C++ interface does not work because vector is always owning and cannot assume ownership of the already initialized memory AFAIK.

Returning a C++ datastructure like vector and use it for np.array is also not possible because you cannot guarantee that the same allocator as in python is used.

You can return an object that wraps the vector and implements the buffer protocol though and then cast to np.array. Also essentially what eigency would do. Or what we do for 1D arrays in autowrap.

Comment (timosachsenberg) [reply src/openms/include/OpenMS/KERNEL/MSExperiment.h:466]: an opaque pointer could probably also work

File: src/openms/source/KERNEL/MSExperiment.cpp [hunk 6]
```diff
@@ -256,6 +260,205 @@ namespace OpenMS
...
+          // Clamp to valid range: values exactly at max_mz should go in last bin
+          if (mz_bin >= 0)
+          {
+            if (mz_bin > mz_bins_minus_one)
+            {
+              mz_bin = mz_bins_minus_one;
+            }
+            const Size pixel_idx = static_cast<Size>(mz_bin) * rt_bins + static_cast<Size>(rt_bin);
+            local_buffer[pixel_idx] += spec[peak_idx].getIntensity();
+          }
```
[the remark this answers is not in the document]
Comment (timosachsenberg) [reply src/openms/source/KERNEL/MSExperiment.cpp:403]: @jpfeuffer true?

**Comment by timosachsenberg** [comment]
@jpfeuffer I added IM support. Can you check the coderabbit comments?

File: src/pyOpenMS/addons/MSExperiment.pyx [hunk 7]
```diff
@@ -1182,3 +1182,87 @@ from libc.stdint cimport uintptr_t
...
+        # Validate array is C-contiguous (required for correct pointer arithmetic)
+        if not output.flags['C_CONTIGUOUS']:
+            raise ValueError("Output array must be C-contiguous. Use np.ascontiguousarray() to convert.")
+
+        cdef _MSExperiment * exp_ = self.inst.get()
+        cdef Size rt_bins = output.shape[1]
+        cdef Size mz_bins = output.shape[0]
+
+        # Get pointer to numpy array data using .data attribute which is safer
+        cdef float* output_ptr = <float*>np.PyArray_DATA(output)
```
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/pyOpenMS/addons/MSExperiment.pyx:1262]: Sure. Why not. Rarely an issue but  safer.

File: src/openms/source/KERNEL/MSExperiment.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpfeuffer) [reply src/openms/source/KERNEL/MSExperiment.cpp:403]: Yeah maybe. Didn't put much thought into the inner optimization. Didn't benchmark also. It just feels so parallelizable.

File: cmake/MacOSX/fix_dependencies.rb (hunk 2 above)
Comment (jpfeuffer) [reply cmake/MacOSX/fix_dependencies.rb:98]: Should be in the commit message. Remove all rpaths but the ones specified. Avoid finding multiple versions of the same lib when copying.
